include/deal.II/optimization/line_minimization.h

```diff
@@ -0,0 +1,620 @@
+
+#ifndef dealii_line_minimization_h
+#define dealii_line_minimization_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/exceptions.h>
+#include <deal.II/base/logstream.h>
+#include <deal.II/base/numbers.h>
+#include <deal.II/base/utilities.h>
+
+#include <deal.II/numerics/history.h>
+
+#include <boost/optional.hpp>
+
+#include <errno.h>
+#include <sys/stat.h>
+
+#include <fstream>
+#include <string>
+
+DEAL_II_NAMESPACE_OPEN
+
+using namespace dealii;
+
+/**
+ * A namespace for various algorithms related to minimization over line.
+ */
+namespace LineMinimization
+{
+  /**
+   * Given $x\_low$ and $x\_hi$ together with values of function
+   * $f(x\_low)$ and $f(x\_hi)$ and the gradient $g(x\_low)$, return the local
+   * minimizer of the quadratic interpolation function.
+   *
+   * The return type is optional to fit with similar function which may
+   * not have solution for given parameters.
+   */
+  template <typename NumberType>
+  boost::optional<NumberType>
+  quadratic_fit(const NumberType x_low,
+                const NumberType f_low,
+                const NumberType g_low,
+                const NumberType x_hi,
+                const NumberType f_hi);
+
+  /**
+   * Given $x\_low$ and $x\_hi$ together with values of function
+   * $f(x\_low)$ and $f(x\_hi)$) and its gradients ($g(x\_low)*g(x\_hi) < 0$) at
+   * those points, return the local minimizer of the cubic interpolation
+   * function. That is the location where the cubic interpolation function
+   * attains its minimum.
+   *
+   * The return type is optional as the real-valued solution might not exist.
+   */
+  template <typename NumberType>
+  boost::optional<NumberType>
+  cubic_fit(const NumberType x_low,
+            const NumberType f_low,
+            const NumberType g_low,
+            const NumberType x_hi,
+            const NumberType f_hi,
+            const NumberType g_hi);
+
+  /**
+   * Find the minimizer of a cubic polynomial that goes through the
+   * points $f\_low=f(x\_low)$, $f\_hi=f(x\_hi)$ and $f\_rec(x\_rec)$
+   * and has derivatve $g\_low$ at $x\_low$.
+   *
+   * The return type is optional as the real-valued solution might not exist.
+   */
+  template <typename NumberType>
+  boost::optional<NumberType>
+  cubic_fit_three_points(const NumberType x_low,
+                         const NumberType f_low,
+                         const NumberType g_low,
+                         const NumberType x_hi,
+                         const NumberType f_hi,
+                         const NumberType x_rec,
+                         const NumberType f_rec);
+
+  /**
+   * Return the minimizer of a polynomial using function values @p f_low @p f_hi @p f_rec[0] at three points
+   * @p x_low @p x_hi @p x_rec[0] and derivatives at two points @p g_low and @p g_hi. The returned point
+   * should be within the bounds @p bounds .
+   *
+   * This function will first try the cubic fit (see cubic_fit ). If it's
```

Comment:
no need for a space before `)`, though you can mark up the function as such by saying `cubic_fit()`.